Python/pystate.c

```diff
@@ -1615,6 +1616,47 @@ PyGILState_Release(PyGILState_STATE oldstate)
         PyEval_SaveThread();
 }
 
+int
+PyThread_TryAcquireFinalizeBlock()
```

Comment:
Please keep this API internal.

Reply:
I am specifically intending to add this as a public API that may be used to safely attempt to acquire the GIL without the risk of hanging the thread (or if we continued to call `pthread_exit`, crashing the program).

The intended use case is as follows:

Previously, you might code like the following:

```python
// some non-Python created thread that wants to send Python an async notification
PyGILState_STATE state = PyGILState_Ensure(); // may kill thread with pthread_exit
// call `call_soon_threadsafe` on some event loop object
PyGILState_Release(state);
```

With this new API:

```python
// some non-Python created thread that wants to send Python an async notification
int acquired = PyThread_TryAcquireFinalizeBlock();
if (!acquired) {
  // skip sending notification since python is exiting
  return;
}
PyGILState_STATE state = PyGILState_Ensure(); // safe now
// call `call_soon_threadsafe` on some event loop object
PyGILState_Release(state);
PyThread_ReleaseFinalizeBlock();
```

Or with the convenience interface:

```python
// some non-Python created thread that wants to send Python an async notification
PyGILState_TRY_STATE state = PyGILState_TryAcquireFinalizeBlockAndGil();
if (!state) {
  // skip sending notification since python is exiting
  return;
}
// call `call_soon_threadsafe` on some event loop object
PyGILState_ReleaseGilAndFinalizeBlock(state);
```

Reply:
This write-up with examples actually makes some sense to be included in your "Cautions regarding runtime finalization" documentation section introducing the new APIs.

Why do we need both APIs.  Why not _only_ offer the "convenience" API as a public API?

Also realize that extension authors and code generators will likely need conditionally compiled code based on the Python version to juse the their existing PyGILState_Ensure and Release calls on <= 3.10 while calling the new convenience API on 3.11+.  When introducing something new, including that in the example should help (though maybe that specific transitional example belongs in the What's New doc for 3.11)

Reply:
In some sense PyThread_TryAcquireFinalizeBlock is logically separate from acquiring the GIL.  Yes, it is hard to imagine why you would want to acquire a finalize block if you don't also want to acquire the GIL, but you may well release the GIL while holding the finalize block (and in fact if the GIL were guaranteed not to be released, then a separate finalize block wouldn't be needed at all).  However, that is a minor point.

Another issue is that the documentation for `PyGILState_Ensure` indicates it should not be used when using sub-interpreters.  I don't really know much about how sub-interpreters work, but as this convenience API is based on `PyGILState_Ensure`, it would have the same limitation.

There is the additional issue of what to do when called from the finalization thread.  Currently these APIs will fail, since they don't make an exception for the finalization thread.  But since you can safely acquire the GIL on the finalization thread, in some cases you might want to skip acquiring the finaliation block, but proceed with acquiring the GIL, if running on the finalization thread.

Reply:
I added this to the documentation, and also added an `#if PY_VERSION_HEX` snippet to the example as well.